### PR TITLE
29 customer fields part 1

### LINKED ISSUE #29: Add fields to customer (I)
Add the following fields to the Customer model.

- Bill To
- Ship To
- Tax Exempt

NB: Federal Tax ID is going to be held at the Entity.

### PR DESCRIPTION
**Pull request by josephbridgwaterrowe**
Closes #29

### CODE REVIEW

**Comment by josephbridgwaterrowe** [comment]
@michaelachrisco please review.

File: .travis.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ script:
 - npm install bower
 - bower install --verbose
 - bundle exec rake db:schema:load
-- bundle exec rspec --color --format documentation
+- xvfb-run bundle exec rspec --color --format documentation
```
Comment (michaelachrisco) [inline .travis.yml]: Interesting, why use xvfb rather than trust bundle exec rspec?

File: app/controllers/customers_controller.rb [hunk 2]
```diff
@@ -39,41 +48,47 @@ def update
...
 
   def customer_hash
     @customer_entry
       .to_h
-      .slice(:default_contact_id,
-             :default_location_id,
-             :is_active,
-             :reference,
-             :salesperson_id)
+      .slice(*param_keys)
```
Comment (michaelachrisco) [inline app/controllers/customers_controller.rb]: I like :+1:

File: app/controllers/customers_controller.rb [hunk 3]
```diff
@@ -39,41 +48,47 @@ def update
...
   def customer_params
     params
       .require(:customer_entry)
-      .permit(:default_contact_id,
-              :default_location_id,
-              :is_active,
-              :reference,
-              :salesperson_id)
+      .permit(
+        param_keys + [
```
Comment (michaelachrisco) [inline app/controllers/customers_controller.rb]: Although this is kind of strange. Is the params optional depending on the customer entry?

File: spec/models/customer_spec.rb [hunk 4]
```diff
@@ -1,26 +1,27 @@
...
-  describe 'validations' do
-    it { is_expected.to validate_presence_of(:entity) }
-    it { is_expected.to validate_presence_of(:is_active) }
-    it { is_expected.to validate_presence_of(:reference) }
-    it { is_expected.to validate_presence_of(:uuid) }
-    it { is_expected.to validate_uniqueness_of(:reference) }
-  end
+  subject { build(:customer, :entity => build(:entity)) }
+
+  it { is_expected.to belong_to(:bill_to_location) }
```
Comment (michaelachrisco) [inline spec/models/customer_spec.rb]: You could add: `belong_to(: bill_to_location).class_name(:location)`, although one can argue if this brings value to the application.

File: spec/models/vendor_spec.rb [hunk 5]
```diff
@@ -33,4 +28,6 @@
       it { is_expected.to eq(true) }
     end
   end
+
+  # TODO: Test long_address and address_lines
```
Comment (michaelachrisco) [inline spec/models/vendor_spec.rb]: Leaving this for the next iteration?

File: spec/support/capybara.rb [hunk 6]
```diff
@@ -1,5 +1,19 @@
 require 'capybara/rails'
 
+Capybara.javascript_driver = :webkit
+
+RSpec.configure do |config|
+  config.before :each, :js, :type => :feature do |example|
+    # Everything is terrible. js: true in config.before will run if the js tag
```
Comment (michaelachrisco) [inline spec/support/capybara.rb]: Interesting....I wonder if there is another driver you can use?

File: app/views/_typeahead.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+$('#<%= value_input %>_typeahead').typeahead({
```
Comment (michaelachrisco) [inline app/views/_typeahead.html.erb]: This looks kind of funky. This is the typeahead render correct? I believe you can add this in as a _typeachead.js.erb (like the example here: http://stackoverflow.com/questions/1127697/js-erb-vs-js)

**Comment by michaelachrisco** [comment]
Added a few comments. Overall looks good.

File: .travis.yml (hunk 1 above)
Comment (josephbridgwaterrowe) [inline .travis.yml]: It was needed to capybara-webkit.

File: spec/support/capybara.rb (hunk 6 above)
Comment (josephbridgwaterrowe) [inline spec/support/capybara.rb]: This one worked, where poltergeist kept shitting the bed.

Comment (michaelachrisco) [inline spec/support/capybara.rb]: Ah ok sounds good. Thanks.

File: spec/models/vendor_spec.rb (hunk 5 above)
Comment (josephbridgwaterrowe) [inline spec/models/vendor_spec.rb]: Yes

File: app/controllers/customers_controller.rb (hunk 3 above)
Comment (josephbridgwaterrowe) [inline app/controllers/customers_controller.rb]: This contains the parameters that are needed between page requests, however the param_keys are only those needed when saving the customer.

File: spec/models/customer_spec.rb (hunk 4 above)
Comment (josephbridgwaterrowe) [inline spec/models/customer_spec.rb]: I think that there is value in including the class_name... I will add in, thanks.

**Comment by josephbridgwaterrowe** [comment]
@michaelachrisco what do you think now?

**Comment by michaelachrisco** [comment]
Looks good. Rebase and I will merge.

**Comment by josephbridgwaterrowe** [comment]
Will do, thanks!

**Comment by josephbridgwaterrowe** [comment]
@michaelachrisco rebased.

**Comment by michaelachrisco** [comment]
:+1:

**Pull request merged**